### ISSUE TITLE
Group tag block is only partially read since 0bba7e1

### ISSUE THREAD

**Issue by imsodin** [dma-ais/AisLib#103]
The properties on `CommentBlockLine` besides the `parameterMap` (e.g. `groupId`) aren't populated anymore since commit "Correct parsing of the messages in comment block parameters that contains G and g" / 0bba7e1. We discovered this when upgrading in a unit test that checked those values on example nmea lines.

I don't understand that commit. The commit message claims it affects lines containing g and G, but it actually acts on comments that start with `\g:`. And the new code seems to be an alernative implementation of the same thing (extract tag keys and values, old code by iteration, new code using `String.split`), with the difference that it skips handling g, G specially, checksum handling and group-id.

From my understanding the old code was working correctly, however I assume something was fixed by that commit - could you please explain what that is.

**Comment by dimitrispallas** [comment]
We had a problem in our system parsing the messages with comment blocks that start with the \g: prefix. With that commit our problems has been resolved. Please feel free to apply a pull request in order to completely fix the issue! Thanks in advance!

**Comment by imsodin** [comment]
Happy to contribute that, however I need to know what exactly was broken for you (I don't see anything obviously wrong with messages starting with \g:) - could you provide some test data that failed and what you'd expect from it (or a unit test)?

**Comment by imsodin** [comment]
I just checked and the unit test that failed with the new version on my side is also a comment block starting with `\g:`, and it was working fine before. So I'll definitely need some example to reproduce your issue.

**Comment by dimitrispallas** [comment]
this is an actual message 24 with comment block that starts with \g:

\g:1-1-21,s:47450,c:1699113260,i:<S>S</S><O>XSP</O><T>A:1699115144 I:1699115146 D:1699115146 L:FM136 G:hndgs T:0 </T><Q>12</Q>*61\!AIVDM,1,1,,A,H7Oks<0hTtpE=<0000000000000,2*4C

Beforehand our system wasn't able to decode and produce the JSON output for that kind of sentence. And we are parsing millions of them in a daily basis. As I already said currently everything works properly in comparison with the previous code.

**Comment by imsodin** [comment]
I got that, I just need to figure out what's wrong to produce a PR that works in both cases - thanks for providing an example for that :)  

I'll probably have a look at this next week.

**Comment by imsodin** [comment]
Obviously that "next week" prediction didn't pan out. It's basically in my backlog now, but given no-one else seems to be affected it anyway is hardly time-critical.

**Comment by mith-emsa** [comment]
Hi @imsodin , did you have the change to pick on the issue again. We would like to understand what could be wrong and try to have the library working for all cases. Thanks

**Comment by FBasels** [comment]
Any update on this? I ran into the same problem that group tags that start the tag block do not set the groupID of the sentence. Looks like a simple fix to me, same for the validation of the checksum. Would be happy to contribute that.
However, I am wondering if the "old" method (`validateCommentBlock`) of parsing is still necessary, or if we could simply apply the new one (`parseCommentBlockWithLetterG`) with some minor changes to every tag block.

**Comment by apostolis-ed** [comment]
Hi all, 
The differences between the functions are the following:

- validateCommentBlock: Splits sentences on both commas (,) and colons (:). As a result, it has undefined behavior when there are multiple colons between 2 commas 
               `e.g.  ,i:<S>A</S><Q>12</Q><O>XLS</O><T>I:1680588456 A:1680588453 G:3895</T>` 

    The outcome of the function depends on the number of commas and colons: 

    1. If there is an odd number of colons between 2 commas, the function will run successfully but the decoded key-value pairs will be malformed. 
     2. If there is an even number of colons between 2 commas, the function will throw an exception.

- parseCommentBlockWithLetterG: Splits sentences on commas only (,) first. Each resulting value is further split by colons (:). If the resulting values are not 2 (key-value pair) they are ignored.

   So in the following example: 

  `\g:1-1-21,s:47450,c:1699113260,i:<S>S</S><O>XSP</O><T>A:1699115144 I:1699115146 D:1699115146 L:FM136 G:hndgs T:0 </T><Q>12</Q>*61\\!AIVDM,1,1,,A,H7Oks<0hTtpE=<0000000000000,2*4C`

    The result would be the following: {s=47450, c=1699113260} (the `i:<S>S</S><O>XSP</O><T>A:1699115144 I:1699115146 D:1699115146 L:FM136 G:hndgs T:0 </T><Q>12</Q>` values are ignored, because if we split by colon (:) the resulting values are more than 2).

    Another difference between the 2 functions is that parseCommentBlockWithLetterG ignores the first tag (e.g. in example above the value g=1-1-21 will be ignored, while validateCommentBlock function would extract such values)

And finally, the checksum, groupId, totalLines and lineNumber values are not populated/handled by the parseCommentBlockWithLetterG function.

If it's agreeable that the parseCommentBlockWithLetterG handles the logic better, we could incorporate whats missing into the function.

**Comment by mith-emsa** [comment]
Hi all,
Providing my input, I believe that we should follow what is stated in the AIS conventions for the comment block and I found this better explained in https://gpsd.gitlab.io/gpsd/AIVDM.html#_nmea_tag_blocks. In particular it is stated that:
“The general format of a tag block is: an opening backslash, followed by multiple comma-separated fields none of which may contain backslashes, followed by an asterisk and NMEA checksum, followed by a closing backslash.”

Therefore I believe that the best approach would be to Splits sentences on commas only (,) first. Then each resulting value would be split by the first colon (ignoring any other colons that could exist) , which would result in a key-value pair for each split done by comma. Inside the value of the pair could be any string. In the case of the example:

`\g:1-1-21,s:47450,c:1699113260,i:<S>S</S><O>XSP</O><T>A:1699115144 I:1699115146 D:1699115146 L:FM136 G:hndgs T:0 </T><Q>12</Q>*61\\!AIVDM,1,1,,A,H7Oks<0hTtpE=<0000000000000,2*4C`

This would lead to `{s=47450, c=1699113260, i=<S>S</S><O>XSP</O><T>A:1699115144 I:1699115146 D:1699115146 L:FM136 G:hndgs T:0 </T><Q>12</Q>}`

Regarding the comment block tag g, this one is used to support multi sentences AIS messages. Most probably the parseCommentBlockWithLetterG  does not store the comment block g, because it uses to group the VDM sentences to be able to decode them and after grouping them it does not have any added value the information in tag g. 

In conclusion I would suggest the following modifications:

1. Deprecate the validateCommentBlock function and replace all comment block processing by  parseCommentBlockWithLetterG
2. Change the logic of processing of parseCommentBlockWithLetterG to consider only the first colon in each comma separated block to create the key-value pairs (which will ensure that there are always 2)
3. Change the parseCommentBlockWithLetterG to support processing comment blocks with g and without g tag
4. Change the parseCommentBlockWithLetterG to populate the checksum, groupId, totalLines and lineNumber values

**Comment by FBasels** [comment]
I just sat down and implemented the suggested modifications. Seems to be fine except for [one test](https://github.com/dma-ais/AisLib/blob/master/ais-lib-communication/src/test/java/dk/dma/ais/reader/AisReaderTest.java#L41) that fails because of the sentence:
``\s:b:2,c:1364272372*57\!AIVDM,1,1,,C,18153ogP?w1dD@@`JiRN4?wp0000,0*48``

In the current implementation, the parsing of the tag is stopped due to the tag parameter `s:b:2` which causes an uneven amount of parameter codes and values. If we change the parsing as discussed in point 2, this would be a valid tag parameter with the value `b:2`.  As there seems to be a need of allowing additional colons (`:`) in parameter values, how shall  that sentence be handled?

**Comment by mith-emsa** [comment]
My suggestion @FBasels is that indeed we change it as mentioned in point 2, leading to the s parameter value having b:2 in it. The s parameter is the identifier of the source station, while b does not seem to be a standard tag used in the comment block. If the string is b:2 it is meaningless but it is also not something completely stupid. It is then for the reader of the decoded message to decide what to do with it. But at least we are able to decode it and not generate complete non-sense. In the end we cannot account for every possible different usage of the tags in the comment block. @FBasels , Thanks for your input on this.

**Comment by apostolis-ed** [comment]
Merged pull request and released version 2.8.6 of the library
